src/rust/catloop/mod.rs

```diff
@@ -428,20 +432,75 @@ impl CatloopLibOS {
                 if let Some(duplex_pipe) = queue.get_pipe() {
                     duplex_pipe.close()?;
                 }
+                // Rollback the port allocation.
                 if let Socket::Active(Some(addr)) | Socket::Passive(addr) = queue.get_socket() {
-                    // Rollback the port allocation
-                    self.ephemeral_ports.borrow_mut().free(addr.port());
+                    if EphemeralPorts::is_private(addr.port()) {
+                        if self.ephemeral_ports.borrow_mut().free(addr.port()).is_err() {
+                            // We fail if and only if we attempted to free a port that was not allocated.
+                            // This is unexpected, but if it happens, issue a warning and keep going,
+                            // otherwise we would leave the queue in a dangling state.
+                            warn!("close(): leaking ephemeral port (port={})", addr.port());
+                        }
+                    }
                 }
                 queue.cancel_pending_ops(Fail::new(libc::ECANCELED, "this queue was closed"));
+                qtable.free(&qd);
+                Ok(())
+            },
+            None => {
+                let cause: String = format!("invalid queue descriptor (qd={:?})", qd);
+                error!("close(): {}", &cause);
+                Err(Fail::new(libc::EBADF, &cause))
+            },
+        }
+    }
+
+    /// Asynchronously closes a socket.
+    pub fn async_close(&mut self, qd: QDesc) -> Result<QToken, Fail> {
+        trace!("async_close() qd={:?}", qd);
+
+        let mut qtable: RefMut<IoQueueTable<CatloopQueue>> = self.qtable.borrow_mut();
+        // Remove socket from sockets table.
+        match qtable.get_mut(&qd) {
+            Some(queue) => {
+                if let Some(duplex_pipe) = queue.get_pipe() {
+                    let qt: QToken = duplex_pipe.async_close()?;
+                    self.catmem_qts.insert(qt, (demi_opcode_t::DEMI_OPC_CLOSE, qd));
+                    Ok(Self::shift_qtoken(qt))
+                } else {
+                    let yielder: Yielder = Yielder::new();
+                    let yielder_handle: YielderHandle = yielder.get_handle();
+                    let coroutine: Pin<Box<Operation>> = Box::pin(async move { (qd, OperationResult::Close) });
```

Comment:
We should free the queue from the qtable here. Not sure how to do it above, I think you might want to use an async block there as well instead of having the duplex pipe insert it into the scheduler.

Reply:
I am not sure what you meant. The `harvest_resources_on_close` will do that in `pack_result`.